torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,731 @@
+from torch.distributed._tensor.redistribute import _redistribute_with_local_tensor
+from torch.fx import GraphModule
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch.fx.passes.shape_prop import _extract_tensor_metadata
+from torch.nn.utils._named_member_accessor import NamedMemberAccessor
+
+
+aten = torch.ops.aten
+
+
+class DataParallelStyle(Enum):
+    """
+    We have three types of Data Parallel style:
+    1. DEFAULT: the default data parallel style, which is to represent a mixed
+                replicate and fully shard behavior. For each parameter that is able
+                to be sharded evenly, we shard it, otherwise we would replicate the
+                parameter. This style avoids potential padding if the parameters
+                cannot be sharded evenly, but it would generate a mixed of all_reduce
+                and reduce_scatter.
+    2. REPLICATE: the data parallel style that replicates all model parameters.
+                  This is similar to the behavior of DistributedDataParallel.
+    3. FULLY_SHARD: the data parallel style that shards all model parameters. This
+                    is similar to the behavior of FullyShardedDataParallel, the
+                    difference is that FullyShardedDataParallel (ZERO-3), which
+                    shards the model using FlatParameter based sharding,
+                    while this style shard each parameter into DTensor.
+    """
+
+    DEFAULT = 0
+    REPLICATE = 1
+    FULLY_SHARD = 2
+
+
+class NodeType(Enum):
+    """
+    NodeType is a enum that records the type of the tensors in the graph.
+    This is used to determine the data parallel strategy.
+    """
+
+    PARAM = 0
+    ACT = 1
+    GRAD = 2
+    STATE = 3
+    NON_TENSOR = 4  # NON_TENSOR is to tag non tensor node (i.e. graph output)
+
+
+@dataclass
+class DataParallelStrategy(StrategyList):
+    """
+    DataParallelStrategy is a special case of StrategyList that only records
+    the "data parallel style" placement strategy for each fx Node.
+    """
+
+    node_type: NodeType
+    reduction_over_batch: bool
+
+    def __init__(
+        self,
+        node_type: NodeType,
+        startegy_list: StrategyList,
+        reduction_over_batch: bool = False,
+    ):
+        super().__init__(startegy_list)
+        self.node_type = node_type
+        self.reduction_over_batch = reduction_over_batch
+
+    def __str__(self) -> str:
+        return f"type: {self.node_type}, {super().__str__()}"
+
+
+class BatchDimAnalyzer(object):
+    """
+    This class is used to analyze the batch dimension of each tensor/node in the
+    graph. We need to know the batch dimension of each tensor/node so that we know
+    exactly the sharding layout of intermediate tensors.
+
+    We possibly should use symbolic shapes to track the batch dimension, but this
+    needs to happen with dynamo, as dynamo is the only place we can mark a single
```

Comment:
could you please elaborate on this one? We can get symints in on the traced graph of local train step, will that be sufficient if we never flatten batch dimension with other dimensions? In some RecSys models, we saw batch dim on the 3rd dimension in the overall input tensor. We can ask users to mark that and track it all the way through symints if necessary.

Curious, why Dynamo can faithfully identify which dim is batch dimension?

Reply:
that would be sufficient, what I am referring to here is that we probably don't want to mark input dimension as sym ints and might only want to mark the batch dimension, currently only dynamo gives this API `mark_dynamic` which allow us to only mark certain tensor dimensions. 

As discussed in the meeting, if symbolic shape in some case can't do cuda graph lowering and drags the performance, we might want to write our own batch dimension analysis in a more canonical way.

Reply:
This claim "dynamo is the only place we can mark a single dimension" is simply untrue.

Here is an example script that does what you want:

```
import torch
from torch.fx.experimental.symbolic_shapes import ShapeEnv, DimDynamic
from torch._subclasses.fake_tensor import FakeTensorMode
        
shape_env = ShapeEnv()
fake_mode = FakeTensorMode(shape_env=shape_env)
        
t = torch.randn(4, 3)
fake_t = fake_mode.from_tensor(t, dynamic_dims=[DimDynamic.DYNAMIC, DimDynamic.STATIC])
print(fake_t.shape)

print(fake_t.sum(dim=0).shape)
print(fake_t.sum(dim=1).shape)
```

this prints

```
torch.Size([s0, 3])
torch.Size([3])
torch.Size([s0])
```

What I will say is that the API surface for direct FakeTensorMode manipulation isn't all that well developed. If you can collect some requirements for us there are probably some easy improvements for your use case.